File: 1/Graph.java
import java.util.ArrayList;

public class Graph<E> {
	ArrayList<Vertex> vertices;	
	
	public Graph() {
		vertices = new ArrayList<Vertex>();
	}
	
	public class Vertex {
		E data;
		ArrayList<Vertex> adjacenctVertices;

		public Vertex(E data) {
			this.adjacenctVertices = new ArrayList<Vertex>();
			this.data = data;
		}

		public void addAdjacentVertex(Vertex to) {
			for(Vertex v : this.adjacenctVertices) {
				if(v.data == to.data) {
					return;
				}
			}
			this.adjacenctVertices.add(to);	
		}	
		
		public E getData() {
			return this.data;
		}

		public ArrayList<Vertex> getAdjacentVertices() {
			return this.adjacenctVertices;
		}
	}
	
	public Vertex getVertex(E data) {
		for(Vertex v : this.vertices ) {
			if(v.data == data) {
				return v;
			}
		}
		return null;
	}
	
	public void addEdge(E from, E to) {
		Vertex fromV, toV;
		fromV = toV = null;

		for(Vertex v : this.vertices) {
			if(v.data == from) {
				fromV = v;
			}
			else if (v.data == to) {
				toV = v;
			}

			if(fromV != null && toV != null) {
				break;
			}
		}

			if(fromV == null) {
				fromV = new Vertex(from);
				this.vertices.add(fromV);
			}
		
			if(toV == null) {
				toV = new Vertex(to);
				this.vertices.add(toV);
			}
			
			fromV.addAdjacentVertex(toV);
	}
	
}

File: 1/js/src/Graph.ts
class Vertex<T> {
	data: T
	children: Vertex<T>[]

	constructor(data: T) {
		this.data = data;
		this.children = [];
	}
}

class Graph<T> {
	vertices: Vertex<T>[]
	
	constructor() {
		this.vertices = [];
	}
	
	getVertexByData(data: T): Vertex<T> | null {
		for(const vertex of this.vertices) {
			if(vertex.data == data) {
				return vertex;
			}
		}
		
		return null;
	}
	 
	addVertex(data: T) {
		const vertex = new Vertex<T>(data);
		this.vertices.push(vertex);

		return vertex;
	}
	
	addEdge(from: T, to: T) {
		let fromVertex = this.getVertexByData(from);	
		if(fromVertex == null) {
			fromVertex = this.addVertex(from);
		}	
		
		let toVertex = this.getVertexByData(to);
		if(toVertex == null) {
			toVertex = this.addVertex(to);
		}

		fromVertex.children.push(toVertex);
	}
}

export {Vertex, Graph};
